src/QsCompiler/Tests.Compiler/TestCases/TypeParameterResolution.qs

```diff
@@ -157,3 +157,110 @@ namespace Microsoft.Quantum.Testing.TypeParameterResolution {
 
 // Foo.A -> Int
 // Bar.B -> Int
+
+// =================================
+
+// Basic Entry Point
+namespace Microsoft.Quantum.Testing.TypeParameterResolution {
+
+    @ EntryPoint()
+    operation Main() : Unit {
+        Foo();
+        Bar();
+    }
+
+    operation Foo() : Unit { }
+
+    operation Bar() : Unit {
+        Baz();
+    }
+
+    operation Baz() : Unit { }
+}
+
+// =================================
+
+// Unrelated To Entry Point
+namespace Microsoft.Quantum.Testing.TypeParameterResolution {
+
+    @ EntryPoint()
+    operation Main() : Unit {
+        Foo();
+    }
+
+    operation Foo() : Unit { }
+
+    operation Bar() : Unit {
+        Baz();
+    }
+
+    operation Baz() : Unit { }
+}
+
+// =================================
+
+// Separated From Entry Point By Specialization
+namespace Microsoft.Quantum.Testing.TypeParameterResolution {
+
+    @ EntryPoint()
+    operation Main() : Unit {
+        Adjoint Foo();
+    }
+
+    operation Foo() : Unit is Adj {
+        body(...) {
+            Bar();
+        }
+
+        adjoint(...) { }
+    }
+
+    operation Bar() : Unit { }
+}
+
+// =================================
+
+// Not Called With Entry Point
+namespace Microsoft.Quantum.Testing.TypeParameterResolution {
+
+    @ EntryPoint()
+    operation Main() : Unit {
+        Foo();
+    }
+
+    operation Foo() : Unit { }
+
+    operation NotCalled() : Unit { }
+}
+
+// =================================
+
+// Not Called Without Entry Point
```

Comment:
I'd group it with the other tests that aren't concerned with entry points. Maybe, there is already coverage for the case of disconnected callgraphs? And, if not, should review what other kinds of callgraph topologies aren't covered (e.g. take "Y" topology, when there are multiple roots that eventually call the same method, is it covered? do we care to test it?).

Reply:
The category for these tests is "Populate Call Graph", making sure that the call graph gets all the expected nodes and edges across a variety of callable relationships, so making sure that unrelated subgraphs are included in the call graph when there is no entry point seemed consistent with that category to me.

Reply:
This comment is similar to my comment in a similar thread on the is PR. The two Not Called tests cover how the call graph walker will respond to an operation that is never called. This is subtly different from a pair of operations that are called but are not related to the rest of the graph. Without an entry point present, these two situations are actually expected to behave differently. A lone node with no relationships should not be present in the call graph.

Reply:
All of the tests prior to this PR can be thought of as "Populate Call Graph in absence of an specified entry point". Consider this collection of tests from the point of view of someone who reads it six months later not as an incremental change but as a whole suite -- does the structure and descriptions of the tests make it clear what is covered and what inputs belong to the same equivalence classes?

---
In reply to: [439693813](https://github.com/microsoft/qsharp-compiler/pull/454#discussion_r439693813) [](ancestors = 439693813)

Reply:
The previous tests in this file are for testing the uses of the GetAllDependencies method of the call graph. This can be seen from the F# code associated with each test case.

Reply:
I don't follow. The tests might be invoked differently but isn't the ultimate goal to validate that the correct callgraph has been created? In which case you systemically list the equivalence classes for the possible callgraphs (including multiple connected components, cycles, isolated nodes, duplicated edges, etc.) and then identify which of these cases would yield new equivalence classes if an entry point is added at different positions inside the graphs... This way you can convince yourself and future readers of the code that the test coverage is good and also document the behaviour.

---
In reply to: [440344069](https://github.com/microsoft/qsharp-compiler/pull/454#discussion_r440344069) [](ancestors = 440344069)

Reply:
This goal of having a correct call graph is only the goal for the tests under the category "Populating Call Graph". It is not the goal of the test in the category "Get Dependencies". The goal of "Get Dependencies" tests is to check the correct behavior of the GetAllDependencies method.